spec/app/ics-100-atomic-swap/README.md

```diff
@@ -0,0 +1,636 @@
+
+#### Cancelling a swap
+
+1.  The maker cancels a previously created order.  Expired orders can also be cancelled.
+2.  An `AtomicSwapPacketData` is relayed to the taker chain where `onRecvPacket` the order is cancelled on the taker chain. If the order is in the process of being taken (a packet with `TakeSwapMsg` is being relayed from the taker chain to the maker chain), the cancellation will be rejected.
+3.  A packet is relayed back where upon acknowledgement the order on the maker chain is also cancelled.  The refund only occurs if the taker chain confirmed the cancellation request.
+
+### Sub-protocols
+
+The sub-protocols described herein should be implemented in a "Fungible Token Swap" module with access to a bank module and to the IBC routing module.
+
+```ts
+function makeSwap(request MakeSwapMsg) {
+  const balance = bank.getBalances(request.makerAddress)
+  abortTransactionUnless(balance.amount > request.sellToken.Amount)
+  // gets escrow address by source port and source channel
+  const escrowAddr = escrowAddress(request.sourcePort, request.sourceChannel)
+  // locks the sellToken to the escrow account
+  const err = bank.sendCoins(request.makerAddress, escrowAddr, request.sellToken)
+  abortTransactionUnless(err === null)
+  // contructs the IBC data packet
+  const packet = {
+    type: SwapMessageType.TYPE_MSG_MAKE_SWAP,
+    data: protobuf.encode(request), // encode the request message to protobuf bytes
+    memo: ""
+  }
+  sendAtomicSwapPacket(packet, request.sourcePort, request.sourceChannel, request.timeoutHeight, request.timeoutTimestamp)
+    
+  // creates and saves order on the maker chain.
+  const order = OrderBook.createOrder(request)
+  //saves order to store
+  store.save(order)
+}
+```
+
+```ts
+function takeSwap(request TakeSwapMsg) {
+  const order = store.findOrderById(request.sourceChannel, request.orderId)
+  abortTransactionUnless(order !== null)
+  abortTransactionUnless(order.expirationTimestamp < currentTimestamp())
+  abortTransactionUnless(order.maker.buyToken.denom === request.sellToken.denom)
+  abortTransactionUnless(order.maker.buyToken.amount === request.sellToken.amount)
+  abortTransactionUnless(order.taker === null)
+  // if `desiredTaker` is set, only the desiredTaker can accept the order.
+  abortTransactionUnless(order.maker.desiredTaker !== null && order.maker.desiredTaker !== request.takerAddress)
+    
+  // check if this take message send to the correct chain
+  abortTransactionUnless(order.channelIdOnTakerChain === request.sourceChannel)
+    
+  const balance = bank.getBalances(request.takerAddress)
+  abortTransactionUnless(balance.amount > request.sellToken.amount)
+  // gets the escrow address by source port and source channel
+  const escrowAddr = escrowAddress(request.sourcePort, request.sourceChannel)
```

Comment:
If I am not mistaken `TakeSwapMsg` does not have a `sourcePort` field, so I guess you need to use `order.portIdOnTakerChain` here instead?

Reply:
yeah.